### PR TITLE
[DCP] Raise RuntimeError on load dtype mismatch to prevent precision …

### LINKED ISSUE #187138: [DCP] load silently casts checkpoint tensors to the template dtype, truncating precision without any warning
### 🐛 Describe the bug

`torch.distributed.checkpoint.load` resolves what to read from the
user-provided `state_dict` template. When a template tensor's dtype differs
from the dtype stored in the checkpoint, the load completes with no warning
or error and the values are silently cast to the template dtype. If the
template dtype is narrower than the checkpoint dtype (for example a
`bfloat16` tensor receiving a `float32` checkpoint value), the restored
state loses precision relative to what was saved, and nothing at any layer
reports it.

The same code path treats a shape mismatch as a hard error, so the silence
is asymmetric: sizes are checked, dtypes are not.

```python
import tempfile

import torch
import torch.distributed.checkpoint as dcp

torch.manual_seed(0)
ckpt = tempfile.mkdtemp()
saved = {"w": torch.rand(4, 3, dtype=torch.float32) * 100}
dcp.save(state_dict=saved, checkpoint_id=ckpt)

template = {"w": torch.zeros(4, 3, dtype=torch.bfloat16)}  # wrong dtype
dcp.load(state_dict=template, checkpoint_id=ckpt)  # no warning, no error

dev = (saved["w"].double() - template["w"].double()).abs().max()
print(f"loaded dtype: {template['w'].dtype} (checkpoint stored float32)")
print(f"max |saved - loaded|: {dev:.6f}")
```

Output (identical on every version listed under Versions below):

```
loaded dtype: torch.bfloat16 (checkpoint stored float32)
max |saved - loaded|: 0.177818
```

## Why this matters in practice

The dangerous direction is the narrowing one. A mixed-precision trainer
that keeps fp32 master weights (or fp32 Adam moments) and resumes through a
template built from a bf16 model restores silently truncated masters: the
job trains on, the loss curve looks plausible, and the high-precision state
the optimizer depended on is gone. Framework history shows this exact class
shipping as silent correctness bugs (for example Megatron-LM PR#2789,
where bf16 working copies desynced from fp32 masters across a load).

## Where it happens

In the filesystem reader's load path, the size mismatch is an explicit
assertion but the dtype is never compared before the copy
(`torch/distributed/checkpoint/filesystem.py`, `_load_tensors` inner
function):

```python
if target_tensor.size() != tensor.size():
    raise AssertionError(
        f"req {req.storage_index} mismatch sizes {target_tensor.size()} vs {tensor.size()}"
    )
target_tensor.copy_(tensor)   # Tensor.copy_ casts dtypes silently
```

The planner already knows the stored dtype
(`item.tensor_data.properties.dtype` in the metadata), so detecting the
mismatch costs nothing at plan time.

## Expected behavior

Either of these would make the behavior safe:

1. raise on dtype mismatch by default, mirroring the size check, with an
   explicit opt-in (planner flag) for intentional load-time casting; or
2. keep the cast but emit a clear warning naming the tensor, the checkpoint
   dtype, and the template dtype.

Neither the `dcp.load` docstring nor `DefaultLoadPlanner`'s documentation
mentions dtype conversion today, so current behavior is also undocumented.

### Versions

Reproduced identically on:

- torch 2.5.1
- torch 2.12.0 (current stable)
- torch 2.13.0.dev20260611 (nightly)

(CPU, single process, `no_dist` path; the copy site is in the shared
filesystem reader, so the distributed path goes through the same code.)

<details><summary>collect_env (torch 2.12.0 stable env)</summary>

```
Collecting environment information...
PyTorch version: 2.12.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.10 (x86_64)
GCC version: (Ubuntu 14.2.0-4ubuntu2) 14.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.40

Python version: 3.12.7 (main, Jun 18 2025, 13:16:51) [GCC 14.2.0] (64-bit runtime)
Python platform: Linux-6.11.0-29-generic-x86_64-with-glibc2.40
Is CUDA available: False
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Ti
Nvidia driver version: 570.195.03
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 7 5800X 8-Core Processor
CPU family:                           25
Model:                                33
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             0
Frequency boost:                      enabled
CPU(s) scaling MHz:                   86%
CPU max MHz:                          4851.0000
CPU min MHz:                          550.0000
BogoMIPS:                             7599.27
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm debug_swap
L1d cache:                            256 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             4 MiB (8 instances)
L3 cache:                             32 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] torch==2.12.0+cpu
[conda] Could not collect
```

</details>

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @LucasLLC @pradeepfn

### PR DESCRIPTION
**Pull request by val3rkq**
### Description
This PR addresses issue #187138.

Currently, `torch.distributed.checkpoint.load` silently casts tensors from the checkpoint to the template's dtype using `Tensor.copy_`. If the template has a lower precision (e.g., `bfloat16`) than the stored checkpoint (e.g., `float32`), precision is lost without any warning or error.

This change introduces an explicit dtype check in `FileSystemReader._load_tensors` before the data is copied. If a mismatch is detected, a `RuntimeError` is raised, mirroring the existing behavior for shape/size mismatches.

Note: I chose `RuntimeError` instead of `AssertionError` for this check as it is more consistent with PyTorch's common practice for user-facing validation errors (similar to dtype checks in standard operators).

### Test Plan
Verified using the reproduction script from #187138.
1. Before fix: Script completes with silent precision loss.
2. After fix: Script correctly raises `RuntimeError` with a descriptive message.

I also ran existing tests that use `read_data` function : `python test/distributed/checkpoint/test_hf_storage.py` and they passed.

Fixes #187138

cc @LucasLLC @pradeepfn

### CODE REVIEW

**Comment by zou3519** [comment]
Not sure who the right reviewer for this is, @wconstab can you route this please?

**Comment by val3rkq** [comment]
@wconstab @RohitRathore1 check pls
